src/adapter/enum_variant.rs

```diff
@@ -0,0 +1,308 @@
+use rustdoc_types::{Item, ItemEnum, Variant};
+use std::borrow::Cow;
+use std::cell::OnceCell;
+use std::fmt;
+use std::num::ParseIntError;
+use std::rc::Rc;
+use std::str::FromStr;
+
+#[non_exhaustive]
+#[derive(Debug, Clone)]
+pub(super) struct EnumVariant<'a> {
+    item: &'a Item,
+    // HACK: to get around closure bounds seen in ./edges.rs:301
+    pub(super) discriminants: Rc<LazyDiscriminants<'a>>,
+    pub(super) index: usize,
+}
+
+#[non_exhaustive]
+#[derive(Debug, Clone)]
+pub(super) struct LazyDiscriminants<'a> {
+    variants: Vec<&'a Variant>,
+    discriminants: OnceCell<Vec<Cow<'a, str>>>,
+}
+
+impl<'a> LazyDiscriminants<'a> {
+    pub(super) fn new(variants: Vec<&'a Variant>) -> Self {
+        Self {
+            variants,
+            discriminants: OnceCell::new(),
+        }
+    }
+
+    pub(super) fn get_discriminants(&self) -> &Vec<Cow<'a, str>> {
+        self.discriminants
+            .get_or_init(|| assign_discriminants(&self.variants))
+    }
+}
+
+impl<'a> EnumVariant<'a> {
+    pub(super) fn new(
+        item: &'a Item,
+        discriminants: Rc<LazyDiscriminants<'a>>,
+        index: usize,
+    ) -> Self {
+        Self {
+            item,
+            discriminants,
+            index,
+        }
+    }
+
+    pub(super) fn variant(&self) -> Option<&'a Variant> {
+        match &self.item.inner {
+            ItemEnum::Variant(v) => Some(v),
+            _ => None,
+        }
+    }
```

Comment:
We probably want to return `&'a Variant` here, and panic in the `None` branch with a suitable message. We should never have a situation where an `EnumVariant` is constructed from a non-`Variant` item.